http3/server.go

```diff
@@ -658,11 +704,46 @@ func (s *Server) Close() error {
 	return err
 }
 
+const shutdownPollIntervalMax = 500 * time.Millisecond
+
 // CloseGracefully shuts down the server gracefully. The server sends a GOAWAY frame first, then waits for either timeout to trigger, or for all running requests to complete.
 // CloseGracefully in combination with ListenAndServe() (instead of Serve()) may race if it is called before a UDP socket is established.
-func (s *Server) CloseGracefully(timeout time.Duration) error {
-	// TODO: implement
-	return nil
+func (s *Server) CloseGracefully(ctx context.Context) error {
+	s.mutex.Lock()
+	s.closed = true
+	if s.closeCancel != nil {
+		s.closeCancel()
+	}
+	s.mutex.Unlock()
+
+	// copied from http.Server Shutdown
+	pollIntervalBase := time.Millisecond
+	nextPollInterval := func() time.Duration {
+		// Add 10% jitter.
+		interval := pollIntervalBase + time.Duration(rand.Intn(int(pollIntervalBase/10)))
+		// Double and clamp for next time.
+		pollIntervalBase *= 2
+		if pollIntervalBase > shutdownPollIntervalMax {
+			pollIntervalBase = shutdownPollIntervalMax
+		}
+		return interval
+	}
+
+	timer := time.NewTimer(nextPollInterval())
+	defer timer.Stop()
+
+	for {
+		if s.connCount.Load() == 0 {
+			return s.Close()
+		}
+		select {
+		case <-ctx.Done():
+			s.Close()
+			return ctx.Err()
+		case <-timer.C:
+			timer.Reset(nextPollInterval())
+		}
+	}
```

Comment:
We could also create a channel that is closed once the number reaches 0. Every Go routine decrementing the counter would check the value of the counter, and close the channel once it's 0. This would allow us to rewrite this as a simple `select` statement, selecting from this channel, and a timer channel.

Reply:
Done